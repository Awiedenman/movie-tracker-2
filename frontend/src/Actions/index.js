export const fetchInitialMovies = movies => ({
  type: 'FETCH_INITIAL_MOVIES',
  movies
});

export const userLogin = user => ({
  type: 'USER_LOGIN',
  id: user.id,
  name: user.name,
  loggedIn: user.loggedIn
});

export const userSignUp = newUserData => ({
  type: 'USER_SIGN_UP',
  id: newUserData.id
});

export const userSignOut = () => ({
  type: 'USER_SIGN_OUT'
});

export const clearUserFavorites = () => ({
  type: 'CLEAR_USER_FAVORITES'
});

export const addFavorite = movie => ({
  type: 'ADD_FAVORITE',
  favorite: movie
});

export const removeFavorite = () => ({
  type: 'REMOVE_FAVORITE'
});

export const addExistingFavorites = existingUserFavorites => {
  return {
    type: 'ADD_EXISTING_FAVORITES',
    favorites : existingUserFavorites
  };
};