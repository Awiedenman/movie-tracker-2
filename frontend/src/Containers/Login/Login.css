.login,
.sign-up {
  border: none;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,.2), 0 1px 1px 0 rgba(0,0,0,.14), 0 2px 1px -1px rgba(0,0,0,.12);
  margin: 2rem auto;
  max-width: 70rem;
  padding: 3rem 2rem;
}

.login__title {
  font-size: 2.5rem;
  margin-bottom: 2rem;
  text-align: center;
}

.login__error {
  color: red;
  font-size: 1.3rem;
  margin-bottom: 2rem;
  text-align: center;
}

.form {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.form__input {
  background-color: #eee;
  border: none;
  border-radius: 0.3rem;
  font-size: 2rem;
  height: 3em;
  margin-bottom: 2rem;
  opacity: 0.5;
  padding: 0.5rem 2rem 0.5rem 1rem;
  transition: opacity .3s ease;
  width: 50rem;
}

.form__input:focus,
.form__input:active,
.form__input:hover {
  opacity: 1;
}

.form__button {
  background: none;
  border: 1px solid salmon;
  border-radius: 0.3rem;
  font-size: 2.2rem;
  padding: 2rem 4rem;
  width: 30rem;
}