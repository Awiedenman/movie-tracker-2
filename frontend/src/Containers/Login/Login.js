import React, { Component } from 'react';
import { connect } from 'react-redux';
import { userLogin } from '../../Actions/index';
import { userLoginRequest } from '../../helpers/apiCalls';
import PropTypes from 'prop-types';

import './Login.css';

export class Login extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      failedLogin: false,
      error: null
    };
  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  handleSubmit = async event => {
    const { email, password } = this.state;
    event.preventDefault();

    try {
      const user = await userLoginRequest(email, password);
      this.props.handleUserLogin(user.data);
      this.setState({ email: '', password: '' });
      this.props.history.push('/');
    } catch (error) {
      this.setState({ failedLogin: true, error, email: '', password: '' });
    }
  }

  render() {
    const { email, password, error, failedLogin } = this.state;

    return (
      <section className="login">
        <h1 className="login__title">Sign In</h1>
        {failedLogin && <p className="login__error">{error.message}</p>}
        <form className="form"  onSubmit={this.handleSubmit}>
          <input type="text"
            value={email}
            name="email"
            placeholder="Email"
            onChange={this.handleChange}
            className="form__input"
            required
          />
          <input
            type="password"
            value={password}
            name="password"
            placeholder="Password"
            onChange={this.handleChange}
            className="form__input"
            required
          />
          <button className="form__button">Sign In</button>
        </form>
      </section>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  handleUserLogin: user => dispatch(userLogin(user))
});

Login.propTypes = {
  handleUserLogin: PropTypes.func
};

export default connect(null, mapDispatchToProps)(Login);