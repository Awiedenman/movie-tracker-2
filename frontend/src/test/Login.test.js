import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapDispatchToProps } from '../Containers/Login/Login';
import { userLogin } from '../Actions/index';

describe('Login', () => {
  let wrapper;

  beforeEach(() => wrapper = shallow(<Login handleUserLogin={jest.fn()}/>));

  test('should update the email prop in state change', () => {
    const mockEvent = {
      target: {
        name: 'email',
        value: 'me@mail.com'
      }
    };

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('email')).toEqual(mockEvent.target.value);
  });

  test('should update the password prop in state change', () => {
    const mockEvent = {
      target: {
        name: 'password',
        value: 'passmethrough'
      }
    };

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('password')).toEqual(mockEvent.target.value);
  });

  test('should update reset state when handleSubmit is invoked', async () => {
    const mockEvent = {
      preventDefault: jest.fn()
    };

    wrapper.setState({
      name: 'myname',
      password: 'pass'
    });

    window.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({
          email: "tman2272@aol.com",
          id: 1,
          name: "Taylor",
          password: "password"
        })
      })
    );

    await wrapper.instance().handleSubmit(mockEvent);
    expect(wrapper.state('password')).toEqual('');
    expect(wrapper.state('email')).toEqual('');
  });

  test('should handle errors when login fails', async () => {
    const mockEvent = {
      preventDefault: jest.fn()
    };

    wrapper.setState({
      name: 'myname',
      password: 'pass'
    });

    window.fetch = jest.fn().mockImplementation(() =>
      Promise.reject({
        ok: false
      }));

    await wrapper.instance().handleSubmit(mockEvent);
    expect(window.fetch).toHaveBeenCalled();
    expect(wrapper.state('failedLogin')).toBe(true);
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mockUser = {
      email: "tman2272@aol.com",
      id: 1,
      name: "Taylor",
      password: "password"
    };

    const actionToDispatch  = userLogin(mockUser);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.handleUserLogin(mockUser);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  test('should match snapshot when hasError', () => {
    wrapper.setState({ hasError: true });
    expect(wrapper).toMatchSnapshot();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
