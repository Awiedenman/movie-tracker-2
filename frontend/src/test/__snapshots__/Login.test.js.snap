// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    handleUserLogin={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Sign In
</h1>,
        <form
          onSubmit={[Function]}
>
          <input
                    name="email"
                    onChange={[Function]}
                    placeholder="email"
                    type="text"
                    value=""
          />
          <input
                    name="password"
                    onChange={[Function]}
                    placeholder="password"
                    type="password"
                    value=""
          />
          <button>
                    Sign In
          </button>
</form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign In",
        },
        "ref": null,
        "rendered": "Sign In",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="text"
              value=""
/>,
            <input
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
/>,
            <button>
              Sign In
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "placeholder": "email",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign In",
            },
            "ref": null,
            "rendered": "Sign In",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Sign In
</h1>,
          <form
            onSubmit={[Function]}
>
            <input
                        name="email"
                        onChange={[Function]}
                        placeholder="email"
                        type="text"
                        value=""
            />
            <input
                        name="password"
                        onChange={[Function]}
                        placeholder="password"
                        type="password"
                        value=""
            />
            <button>
                        Sign In
            </button>
</form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign In",
          },
          "ref": null,
          "rendered": "Sign In",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                name="email"
                onChange={[Function]}
                placeholder="email"
                type="text"
                value=""
/>,
              <input
                name="password"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
/>,
              <button>
                Sign In
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "email",
                "onChange": [Function],
                "placeholder": "email",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "password",
                "onChange": [Function],
                "placeholder": "password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sign In",
              },
              "ref": null,
              "rendered": "Sign In",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
